sound/soc/sof/intel/hda-stream.c

```diff
@@ -119,13 +119,33 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
 	int remain, ioc;
 
 	period_bytes = hstream->period_bytes;
-	dev_dbg(sdev->dev, "period_bytes:0x%x\n", period_bytes);
-	if (!period_bytes)
+	dev_dbg(sdev->dev, "period_bytes: %#x, bufsize: %#x\n", period_bytes,
+		hstream->bufsize);
+
+	/*
+	 * HDA spec demands at least two 'periods' for the DMA to work correctly
+	 * Note: period_bytes == 0 can only happen for firmware or library
+	 * loading, the binary sizes have the necessary alignment requirements
+	 * met
+	 */
+	if (!period_bytes) {
+		unsigned int chunk_size;
+
+		chunk_size = snd_sgbuf_get_chunk_size(dmab, 0, hstream->bufsize);
+
 		period_bytes = hstream->bufsize;
+		/*
+		 * If the memory is a single continuous area, force it to be
+		 * split into two 'periods', otherwise the transfer will be
+		 * split to multiple BDLE for each chunk in hda_setup_bdle()
+		 */
+		if (chunk_size == hstream->bufsize)
+			period_bytes /= 2;
+	}
 
 	periods = hstream->bufsize / period_bytes;
 
-	dev_dbg(sdev->dev, "periods:%d\n", periods);
+	dev_dbg(sdev->dev, "periods: %d\n", periods);
```

Comment:
The commit message says 

"
While the LVI=0 worked so far without issues, it is better to align with
the specification to avoid unforeseen issues with FW loading.
"

But there are no issues with FW loading that are fixed by this commit.

Reply:
That is true, LVI=0 worked correctly but the specification explicitly saying that LVI must be at least 1 for the DMA to start.
We don't really know why LVI=0 but it cannot be take granted as it is against the spec.
It could be that SPIB changes the rules to allow LVI=0, but it is not documented afaik.

Reply:
SPIB will definitively prevent the DMA from continuing, so and since we use SPIB for firmware download I really don't see what this fixes.

We've been using this for over 10 years now, if it was wrong we would have detected it, no?

I think the risk of splitting firmware downloads in two chunks offsets the unproven benefit of fixing a hypothetical problem.

wow this almost sounds like fridge poetry haha.